src/computability/tm_computable.lean

```diff
@@ -0,0 +1,220 @@
+/-
+Copyright (c) 2020 Pim Spelier, Daan van Gent. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pim Spelier, Daan van Gent.
+-/
+
+import computability.encoding
+import computability.turing_machine
+import data.polynomial.basic
+import data.polynomial.eval
+
+/-!
+# Computable functions
+
+This file contains the definition of a Turing machine with some finiteness conditions
+(bundling the definition of TM2 in turing_machine.lean), a definition of when a TM gives a certain
+output (in a certain time), and the definition of computability (in polytime or any time function)
+of a function between two types that have an encoding (as in encoding.lean).
+
+## Main theorems
+
+- `id_computable_in_poly_time` : a TM + a proof it computes the identity on a type in polytime.
+- `id_computable`              : a TM + a proof it computes the identity on a type.
+
+## Implementation notes
+
+To count the execution time of a Turing machine, we have decided to count the number of times the
+`step` function is used. Each step executes a statement (of type stmt); this is a function, and
+generally contains multiple "fundamental" steps (pushing, popping, so on). However, as functions
+only contain a finite number of executions and each one is executed at most once, this execution
+time is up to multiplication by a constant the amount of fundamental steps.
+-/
+
+open computability
+
+/-- A bundled Turing Machine of type 2 (as defined in turing_machine.lean), with an input and
```

Comment:
Could you clarify the reference, maybe linking to a particular declaration? (Putting a full identifier, e.g. `turing_machine.type_2`, in backicks will get hyperlinked in the docs.) A quick search in `turing_machine.lean` doesn't turn up anything about "type 2," it took some reading and guesswork to find `TM2`.

Reply:
Thanks, a specific reference is indeed better! Now I refer to a namespace in a certain file, will that get hyperlinked nicely? (There is no one definition of a TM2 to refer to)

Reply:
Namespaces in Lean don't belong to specific files. So they can't get hyperlinked. (You can add new stuff to the `polynomial` namespace in 20 different files.) If there is not *one* definition, then maybe just list two or three. Docstrings can't really become "too long".

Reply:
With "there is no one definition" I meant; the definition is spread out over the file (one could define a TM as a function `M : \Lambda \to stmt`, but that already presupposes the number of stacks, the alphabets, and the names of the different functions). One could see a TM as the collection of the number of stacks, the alphabets, the function names and the function definitions, but they're all floating around as parameters, so we can't refer to that either. And referring to a certain line in the file sounds like a bad idea too.

If somebody comes up with a better way to refer than the current "defined starting from the namespace `turing.TM2` in `turing_machine.lean`)", I'd be happy to hear it, but we'll keep it as is for now.

Reply:
I call it "the TM2 model" in documentation in that file, if it helps.

Reply:
As for linking to the docs, ideally it would go to the module doc comment entitled "The TM2 model" but unfortunately there is no anchor for that in the generated docs (@robertylewis ?). The next best option is [`turing.TM2.stmt`](https://leanprover-community.github.io/mathlib_docs/computability/turing_machine.html#turing.TM2.stmt), the first declaration in the section (also the most useful to look at to understand the model).

Reply:
The docs do generate an anchor called `#top`, but the only link to it is in the sidebar. Actually, it should be possible to give all the markdown headings anchors (with `#` that link to them like on our website). We should open an issue at `doc-gen` for that and I'll try to tackle it when I have time.

Reply:
In the absence of anchor links, this seems good to me!